Userland/Libraries/LibJS/Runtime/Intl/CollatorCompareFunction.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2022, Idan Horowitz <idan.horowitz@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Utf16View.h>
+#include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/Intl/Collator.h>
+#include <LibJS/Runtime/Intl/CollatorCompareFunction.h>
+
+namespace JS::Intl {
+
+CollatorCompareFunction* CollatorCompareFunction::create(GlobalObject& global_object, Collator& collator)
+{
+    return global_object.heap().allocate<CollatorCompareFunction>(global_object, global_object, collator);
+}
+
+CollatorCompareFunction::CollatorCompareFunction(GlobalObject& global_object, Collator& collator)
+    : NativeFunction(*global_object.function_prototype())
+    , m_collator(collator)
+{
+}
+
+void CollatorCompareFunction::initialize(GlobalObject& global_object)
+{
+    auto& vm = global_object.vm();
+    define_direct_property(vm.names.length, Value(2), Attribute::Configurable);
+}
+
+double compare_strings(Collator& collator, Utf16View const& x, Utf16View const& y)
+{
+    // FIXME: Implement https://unicode.org/reports/tr10
+    (void)collator;
+    auto x_iterator = x.begin();
+    auto y_iterator = y.begin();
+    for (; x_iterator != x.end() && y_iterator != y.end(); ++x_iterator, ++y_iterator) {
+        if (*x_iterator != *y_iterator)
+            return static_cast<double>(*x_iterator) - static_cast<double>(*y_iterator);
+    }
+    if (x_iterator != x.end())
+        return 1.0;
+    if (y_iterator != y.end())
+        return -1.0;
+    return 0.0;
+}
+
+// 10.3.3.1 Collator Compare Functions, https://tc39.es/ecma402/#sec-collator-compare-functions
+ThrowCompletionOr<Value> CollatorCompareFunction::call()
+{
+    auto& vm = this->vm();
+    auto& global_object = this->global_object();
+
+    // 1. Let collator be F.[[Collator]].
+    // 2. Assert: Type(collator) is Object and collator has an [[InitializedCollator]] internal slot.
+    // 3. If x is not provided, let x be undefined.
+    // 4. If y is not provided, let y be undefined.
+
+    // 5. Let X be ? ToString(x).
+    auto x = TRY(vm.argument(0).to_utf16_string(global_object));
+    // 6. Let Y be ? ToString(y).
+    auto y = TRY(vm.argument(1).to_utf16_string(global_object));
```

Comment:
Do these need to be UTF-16? The spec says "String values must be interpreted as UTF-16 code unit sequences, and a surrogate pair ... within a string must be interpreted as the corresponding code point." So that sounds like we can just use `to_string()` + `Utf8View` to extract the code points without transcoding

Reply:
Nope, I assumed we would since Segmenter did, but from skimming the TR10 spec it doesn't seem like we will.